### ISSUE TITLE
[react/prop-types] False positive when destructuring

### ISSUE THREAD

**Issue by zimme** [yannickcr/eslint-plugin-react#928]
Hey

I've run into a false positive on `react/prop-types` when doing destructuring.

I've created a branch with the failing test-case [here](https://github.com/zimme/eslint-plugin-react/tree/prop-types/false-positive).
The test-case is as close to what we had when it triggered.

``` js
}, {
  code: [
    'class Hello extends React.Component {',
    '  props: {',
    '    name: {',
    '      firstname: string;',
    '    }',
    '  };',
    '  arbitrary({one, two}: {one: string, two: string}) {}', // @flow annotation
    '  render () {',
    '    const { lastname } = this.props.name;',
    '    return <div>Hello {lastname}</div>;',
    '  }',
    '}'
  ].join('\n'),
  parser: 'babel-eslint',
  errors: [
    {message: '\'name\' is missing in props validation'}
  ]
}
```

**Comment by kurtzur** [comment]
I also ran into this today. The code that triggered the false positive was like so:

```js
function displaySomething({ startDate, endDate }) { //offending line
  // ...
}

const MyComponent = React.createClass({
  propTypes: {
    startDate: React.PropTypes.string.isRequired,
    endDate: React.PropTypes.string.isRequired,
  },

  render() {
    return <div>{ displaySomething(this.props) }</div>;
  }
});
```

**Comment by ljharb** [comment]
@kurtzur does `displaySomething` return JSX? if so, our component detection probably things it's an SFC (and it probably should be one)

**Comment by zimme** [comment]
@ljharb,
In his specific case it seems like it should be a SFC as he's providing `this.props` to the function, but what if the function returns JSX because it's a functions that's gonna be used in a `map`?
```
render () {
  return (<div>{ this.props.items.map(displaySomething) }</div>);
}
```

**Comment by ljharb** [comment]
@zimme `return (<div>{this.props.items.map(item => (<SFC item={item} />))}</div>);`

**Comment by Andreyco** [comment]
I got false positive connected to destruction assignment.

Code preview
```jsx
export default class NavigationButton extends React.Component {
  static propTypes = {
    route: PropTypes.shape({
      getBarTintColor: PropTypes.func.isRequired,
    }).isRequired,
  };

  renderTitle() {
    const { route } = this.props;
    return <Title tintColor={route.getBarTintColor()}>TITLE</Title>;
  }
}
```

I get this false report, despite `getBarTintColor` is clearly used.
```
'route.getBarTintColor'PropType is defined but prop is never used
```

**Comment by yannickcr** [comment]
@zimme Thanks for your test case, but when I run it with eslint-plugin-react 6.5.0 I get `'name.lastname' is missing in props validation`, which see correct to me (in your component you declare `name.firstname` but not `name.lastname`). Did I miss something ?

@kurtzur Like @ljharb said, if `displaySomething` is returning JSX then we considere it as an SFC and you must declare propTypes on it.

@Andreyco Seems your talking about the `react/no-unused-prop-types` rule, this issue is about the `react/prop-types`. `react/no-unused-prop-types` has currently some issues and a lot of false positive :cry:

**Comment by zimme** [comment]
Either this has been fixed since I ran into the problem or I just messed something up with the test case. Anyways I can't really seem to replicate the problem and I've since changed my code so I don't have this problem at the moment.

Maybe it was related to @kurtzur problem which I understand why it would seem like a false positive but isn't.

I'll just open another issue if I run into this again.

**Comment by sb8244** [comment]
I see this is recently closed, but wanted to chime in that I am seeing this problem as well. I'm going to go the route of very small SFC even though I'm not entirely convinced that is a solution to all problems.
